### PR TITLE
ENG-4686: Added basic liquibase capability to multitenant

### PR DESCRIPTION
**Pull request by eugeniosant**

### CODE REVIEW

File: engine/src/main/java/org/entando/entando/aps/servlet/StartupListener.java [hunk 1]
```diff
@@ -73,17 +74,18 @@ public void contextInitialized(ServletContextEvent event) {
...
         ITenantManager tenantManager = ApsWebApplicationUtils.getBean(ITenantManager.class, svCtx);
+        IInitializerManager initManager = ApsWebApplicationUtils.getBean(IInitializerManager.class, svCtx);
         tenantManager.getCodes().stream().forEach(tenantCode -> {
-
             ApsTenantApplicationUtils.setTenant(tenantCode);
             try {
                 LOGGER.info("refresh bean for tenant:'{}'", tenantCode);
+                initManager.init();
                 ApsWebApplicationUtils.executeSystemRefresh(svCtx);
+                String tenantMsg = this.getClass().getName() + ": Tenant '" + tenantCode + "' inizialized";
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/servlet/StartupListener.java]: you should use String.format and avoid to do the work of slf4j (class name ...)

File: engine/src/main/java/org/entando/entando/aps/system/init/AbstractInitializerManager.java [hunk 2]
```diff
@@ -33,18 +33,23 @@ public abstract class AbstractInitializerManager implements BeanFactoryAware {
...
+			InstallationReportDAO dao = this.createReportDAO();
 			report = dao.loadReport(this.getConfigVersion());
 		} catch (Throwable t) {
 			_logger.error("error Error extracting report", t);
 			throw new EntException("Error extracting report", t);
 		}
 		return report;
 	}
+    
+    protected InstallationReportDAO createReportDAO() {
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/init/AbstractInitializerManager.java:45]: why this method? who overrides it? we should prefer composition over inheritance ...

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java [hunk 3]
```diff
@@ -93,26 +99,28 @@ public class DatabaseManager extends AbstractInitializerManager
     private int lockFallbackMinutes;
 
     private ServletContext servletContext;
+    private ITenantManager tenantManager;
 
     public void init() {
         logger.debug("{} ready", this.getClass().getName());
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: pls change it

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java [hunk 4]
```diff
@@ -93,26 +99,28 @@ public class DatabaseManager extends AbstractInitializerManager
...
     }
-
+    
     @Override
-    public SystemInstallationReport installDatabase(SystemInstallationReport report, DatabaseMigrationStrategy migrationStrategy) throws Exception {
+    public SystemInstallationReport installDatabase(SystemInstallationReport report, DatabaseMigrationStrategy defaultMigrationStrategy) throws Exception {
         String lastLocalBackupFolder = null;
-        migrationStrategy = (null == migrationStrategy) ? DatabaseMigrationStrategy.DISABLED : migrationStrategy;
+        DatabaseMigrationStrategy strategy = Optional.ofNullable(this.getTenantMigrationStrategy())
+                .orElse(Optional.ofNullable(defaultMigrationStrategy).orElse(DatabaseMigrationStrategy.DISABLED));
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: this is not readable and optional inside optional is not a good practice ...
you should move the second on top and change the return type of the new method `getTenantMigrationStrategy` to Optional<String> (Optional was born to use it in the return of signature)

something like this:
```
DatabaseMigrationStrategy defaultComputedStrategy = Optional.ofNullable(defaultMigrationStrategy).orElse(DatabaseMigrationStrategy.DISABLED);
DatabaseMigrationStrategy strategy = getTenantMigrationStrategy().orElse(defaultComputedStrategy);
```

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java [hunk 5]
```diff
@@ -93,26 +99,28 @@ public class DatabaseManager extends AbstractInitializerManager
...
         }
 
         // Check if we are dealing with an old database version (not Liquibase compliant - Entando <= 6.3.2)
         legacyDatabaseCheck();
 
         try {
-            initComponents(report, migrationStrategy);
-            if (DatabaseMigrationStrategy.AUTO.equals(migrationStrategy) && report.getStatus()
+            initComponents(report, strategy);
+            if (DatabaseMigrationStrategy.AUTO.equals(strategy) && report.getStatus()
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: when you compare the constants always before var ...

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java [hunk 6]
```diff
@@ -135,6 +143,20 @@ public SystemInstallationReport installDatabase(SystemInstallationReport report,
         return report;
     }
 
+    private DatabaseMigrationStrategy getTenantMigrationStrategy() {
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: here there are things to improve ... an example ...
```
    private Optional<DatabaseMigrationStrategy> getTenantMigrationStrategy() {
        try {
            return ApsTenantApplicationUtils.getTenant()
                    .map(getTenantManager()::getConfig)
                    .map(TenantConfig::getDbMigrationStrategy)
                    .filter(StringUtils::isNotBlank)
                    .map(String::toUpperCase)
                    .map(DatabaseMigrationStrategy::valueOf);
        } catch (IllegalArgumentException ex) {
            logger.warn("Error to retrieve a valid MigrationStrategy from current tenant",ex);
            return  Optional.empty();
        }
    }

```

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java [hunk 7]
```diff
@@ -170,20 +192,19 @@ private void initComponents(SystemInstallationReport report, DatabaseMigrationSt
     }
 
     private void legacyDatabaseCheck() throws DatabaseMigrationException, SQLException {
-
-        for (DataSource dataSource : defaultDataSources) {
-
+        Optional<DataSource> tenantDs = ApsTenantApplicationUtils.getTenant()
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: usually `isPresent` is used only in `if` statement and not if at the end you should do a map operation ...

```
        List<DataSource> dss = ApsTenantApplicationUtils.getTenant()
                                .map(getTenantManager()::getDatasource)
                                .map(ds -> Arrays.asList(ds))
                                .orElse(defaultDataSources());
```

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java [hunk 8]
```diff
@@ -302,6 +322,12 @@ private List<ChangeSetStatus> executeLiquibaseUpdate(Date timestamp, String comp
         }
         return changeSetToExecute;
     }
+    
+    private DataSource getRightDatasource(String dataSourceName) {
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: `getRightDatasource` as no meaning ... something like `fetchFromTenantOrGetDefaultByName`
and here ( in general in the new method to fill the lack of logs) pls put a debug log to understand what's happening

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java [hunk 9]
```diff
@@ -635,4 +661,13 @@ public void setDefaultDataSources(List<DataSource> defaultDataSources) {
     public void setLockFallbackMinutes(int lockFallbackMinutes) {
         this.lockFallbackMinutes = lockFallbackMinutes;
     }
+
+    protected ITenantManager getTenantManager() {
+        return tenantManager;
+    }
+    @Autowired
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java:679]: is it possible to move the Autowire in the constructor?  .... it's the best practice ...

File: engine/src/main/java/org/entando/entando/aps/system/init/IInitializerManager.java [hunk 10]
```diff
@@ -21,6 +21,8 @@
 public interface IInitializerManager {
 
 	public static enum DatabaseMigrationStrategy {DISABLED, AUTO, GENERATE_SQL};
+    
+    public void init() throws Exception;
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/init/IInitializerManager.java]: why we need to expose an `init` method ? usually if you need something to do after you should use the `InitializingBean` interface ...

File: engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantConfig.java [hunk 11]
```diff
@@ -122,6 +123,10 @@ public String getDbPassword() {
         return configs.get(DB_PASSWORD_PROPERTY);
     }
 
+    public String getDbMigrationStrategy() {
```
Comment (entgigi) [inline engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantConfig.java]: here, if you need, we could think to return an Optional, this has 2 pros:
- inform consumer about the possibility to empty value
- reduce the code to check

File: engine/src/test/java/org/entando/entando/aps/system/init/DatabaseManagerTest.java [hunk 12]
```diff
@@ -297,7 +301,52 @@ private ListableBeanFactory getMockedBeanFactory() {
         Mockito.when(beanFactory.getBean("portDataSource")).thenReturn(Mockito.mock(DataSource.class));
         return beanFactory;
     }
-
+    
+    @Test
+    void testMigrationStrategyForTenant_1() throws Exception {
```
Comment (entgigi) [inline engine/src/test/java/org/entando/entando/aps/system/init/DatabaseManagerTest.java]: good practice says: test method names are **very important**
an example (first i found in a fast search about this golden rule) https://medium.com/@stefanovskyi/unit-test-naming-conventions-dd9208eadbea

**Review by entgigi — CHANGES_REQUESTED** [review]

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java (hunk 7 above)
Comment (entgigi) [reply engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: my bad ... an error:
`.orElse(defaultDataSources());`
should be:
`.orElseGet(() -> defaultDataSources());`

File: engine/src/main/java/org/entando/entando/aps/servlet/StartupListener.java (hunk 1 above)
Comment (eugeniosant) [reply engine/src/main/java/org/entando/entando/aps/servlet/StartupListener.java]: Done

File: engine/src/main/java/org/entando/entando/aps/system/init/AbstractInitializerManager.java (hunk 2 above)
Comment (eugeniosant) [reply engine/src/main/java/org/entando/entando/aps/system/init/AbstractInitializerManager.java:45]: This method is used also in InitializerManager.java

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java (hunk 3 above)
Comment (eugeniosant) [reply engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: Changed

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java (hunk 4 above)
Comment (eugeniosant) [reply engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: Upgraded

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java (hunk 8 above)
Comment (eugeniosant) [reply engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java]: Done

File: engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java (hunk 9 above)
Comment (eugeniosant) [reply engine/src/main/java/org/entando/entando/aps/system/init/DatabaseManager.java:679]: Removed Autowired

**Review by entgigi — APPROVED** [review]
LGTM

**Pull request merged**